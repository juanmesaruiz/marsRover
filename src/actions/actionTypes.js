export const ROVER_CLEAR = 'ROVER@CLEAR';
export const ROVER_CLEAR_POSITION_LOG = 'ROVER@CLEAR_POSITION_LOG';
export const ROVER_DELETE_OBSTACLE = 'ROVER@DELETE_OBSTACLE';
export const ROVER_NEW_INSTRUCTIONS = 'ROVER@NEW_INSTRUCTIONS';
export const ROVER_NEW_MOVE = 'ROVER@NEW_MOVE';
export const ROVER_SET_GRID = 'ROVER@SET_GRID';
export const ROVER_SET_OBSTACLES = 'ROVER@SET_OBSTACLES';
export const ROVER_SET_POSITION = 'ROVER@SET_POSITION';
