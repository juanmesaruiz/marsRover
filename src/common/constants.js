export const ROVER_DIRECTION = {
  N: 'north',
  S: 'south',
  E: 'east',
  W: 'west',
};

export const ROVER_MOVEMENT = {
  F: 'front',
  L: 'left',
  R: 'right',
};

export const KEYBOARDS_CODES = {
  A: 'KeyA',
  D: 'KeyD',
  W: 'KeyW',
};
