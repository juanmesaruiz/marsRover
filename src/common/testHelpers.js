import React from 'react';

export const mockComponent = (name, props) => {
  const { children, ...restOfProps } = props;
  return (
    <code name={name} aria-label={name} props={restOfProps}>
      {children}
    </code>
  );
};
