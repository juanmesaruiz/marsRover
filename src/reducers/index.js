import { combineReducers } from 'redux';

import grid from './grid';
import obstacles from './obstacles';
import rover from './rover';

const reducers = {
  grid,
  obstacles,
  rover,
};

export const getGrid = state => state.grid;
export const getObstacles = state => state.obstacles;
export const getRover = state => state.rover;

export default combineReducers(reducers);
