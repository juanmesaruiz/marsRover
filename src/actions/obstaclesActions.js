import { ROVER_SET_OBSTACLES } from './actionTypes';

export const createObstacles = obstacles => ({
  type: ROVER_SET_OBSTACLES,
  payload: {
    obstacles,
  },
});
