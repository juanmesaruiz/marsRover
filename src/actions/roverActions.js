import { ROVER_CLEAR, ROVER_NEW_MOVE, ROVER_SET_POSITION } from './actionTypes';

export const roverClear = () => ({
  type: ROVER_CLEAR,
});

export const roverSetPosition = position => ({
  type: ROVER_SET_POSITION,
  payload: {
    position,
  },
});

export const roverNewKeyboardMove = ({ code, grid, obstaclesCoordinates }) => ({
  type: ROVER_NEW_MOVE,
  payload: {
    code,
    grid,
    obstaclesCoordinates,
  },
});
