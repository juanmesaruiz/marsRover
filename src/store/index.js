import { applyMiddleware, compose, createStore } from 'redux';
import thunk from 'redux-thunk';

import reducers from '../reducers';

const configureStore = () => {
  const initialState = {};
  const enhancers = [];
  const devToolsExtension = window.__REDUX_DEVTOOLS_EXTENSION__;

  if (typeof devToolsExtension === 'function') {
    enhancers.push(devToolsExtension({ name: '- ROVER MARS -' }));
  }

  const composedEnhancers = compose(applyMiddleware(thunk), ...enhancers);

  const store = createStore(reducers, initialState, composedEnhancers);

  return { store };
};

export default configureStore;
