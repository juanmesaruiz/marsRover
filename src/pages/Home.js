import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import {
  roverClear,
  roverNewKeyboardMove,
  roverSetPosition,
} from '../actions/roverActions';
import { createObstacles } from '../actions/obstaclesActions';
import { getGrid, getObstacles } from '../reducers';

import Actions from '../components/Actions';
import Description from '../components/Description';
import Grid from '../components/Grid';
import LogPosition from '../components/LogPosition';
import Title from '../components/Title';

import {
  getRandomCoordinates,
  getRandomRoverPosition,
} from '../common/helpers';
import { INITIAL_OBSTACLES, KEYBOARDS_CODES } from '../common/constants';

const Home = ({
  className,
  createObstacles,
  grid,
  obstaclesCoordinates,
  roverClear,
  roverNewKeyboardMove,
  roverSetPosition,
}) => {
  useEffect(() => {
    const obstaclesCoordinates = getRandomCoordinates(grid, INITIAL_OBSTACLES);
    const roverPosition = getRandomRoverPosition(grid, obstaclesCoordinates);
    createObstacles(obstaclesCoordinates);
    roverSetPosition(roverPosition);
  }, [createObstacles, grid, roverSetPosition]);

  useEffect(() => {
    const eventListener = event => {
      const { code } = event;
      switch (code) {
        case KEYBOARDS_CODES.A:
        case KEYBOARDS_CODES.D:
        case KEYBOARDS_CODES.W:
          return roverNewKeyboardMove({ code, grid, obstaclesCoordinates });
        default:
          return null;
      }
    };
    document.addEventListener('keydown', eventListener);
    return () => document.removeEventListener('keydown', eventListener);
  }, [grid, obstaclesCoordinates, roverNewKeyboardMove]);

  useEffect(() => {
    return () => {
      roverClear();
    };
  }, [roverClear]);

  return (
    <div className={className}>
      <header>
        <Title />
      </header>
      <section>
        <Description />
      </section>
      <main className="main-container">
        <Grid />
        <LogPosition />
        <Actions />
      </main>
    </div>
  );
};

const StyledHome = styled(Home)`
  padding: 50px;
  text-align: center;

  .main-container {
    margin-top: 20px;
  }
`;

const mapStateToProps = state => ({
  grid: getGrid(state),
  obstaclesCoordinates: getObstacles(state),
});

export default connect(mapStateToProps, {
  createObstacles,
  roverClear,
  roverNewKeyboardMove,
  roverSetPosition,
})(StyledHome);
