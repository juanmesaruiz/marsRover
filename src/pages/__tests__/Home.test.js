import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';

import {
  roverClear,
  roverNewKeyboardMove,
  roverSetPosition,
} from '../../actions/roverActions';
import { createObstacles } from '../../actions/obstaclesActions';
import { getGrid, getObstacles } from '../../reducers';

import { mockComponent } from '../../common/testHelpers';
import { KEYBOARDS_CODES } from '../../common/constants';

import store from '../../store/__mocks__/mockStore';

import Home from '../Home';

jest.mock('../../actions/roverActions');
jest.mock('../../actions/obstaclesActions');
jest.mock('../../reducers/index');

jest.mock('../../components/Actions', () => props =>
  mockComponent('Actions', props)
);
jest.mock('../../components/Description', () => props =>
  mockComponent('Description', props)
);
jest.mock('../../components/Grid', () => props => mockComponent('Grid', props));
jest.mock('../../components/LogPosition', () => props =>
  mockComponent('LogPosition', props)
);
jest.mock('../../components/Title', () => props =>
  mockComponent('Title', props)
);

const mockCoordinates = [
  { x: 2, y: 2 },
  { x: 3, y: 3 },
];
const mockGrid = { x: 50, y: 75 };
const mockObstaclesCoordinates = [{ x: 1, y: 1 }];
const mockRoverPosition = { x: 5, y: 5 };

jest.mock('../../common/helpers', () => ({
  ...jest.requireActual('../../common/helpers'),
  getRandomCoordinates: () => mockCoordinates,
  getRandomRoverPosition: () => mockRoverPosition,
}));

describe('Home test suite', () => {
  const component = (
    <Provider store={store}>
      <Home />
    </Provider>
  );

  const setup = () => render(component);

  beforeEach(() => {
    roverClear.mockReturnValue({
      type: 'homeClear',
    });
    roverNewKeyboardMove.mockReturnValue({
      type: 'roverNewKeyboardMove',
    });
    roverSetPosition.mockReturnValue({
      type: 'roverSetPosition',
    });
    createObstacles.mockReturnValue({
      type: 'createObstacles',
    });
    getGrid.mockReturnValue(mockGrid);
    getObstacles.mockReturnValue(mockObstaclesCoordinates);
  });

  afterEach(jest.clearAllMocks);

  it('Renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(component, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Should match snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('Should create obstacles and set rover position', () => {
    setup();
    expect(createObstacles).toHaveBeenCalledTimes(1);
    expect(createObstacles).toHaveBeenCalledWith(mockCoordinates);
    expect(roverSetPosition).toHaveBeenCalledTimes(1);
    expect(roverSetPosition).toHaveBeenCalledWith(mockRoverPosition);
  });

  it('Should fire new keyboardMove on press A', () => {
    const { getByLabelText } = setup();
    fireEvent.keyDown(getByLabelText('Grid'), { key: 'A', code: 'KeyA' });
    expect(roverNewKeyboardMove).toHaveBeenCalledTimes(1);
    expect(roverNewKeyboardMove).toHaveBeenCalledWith({
      code: KEYBOARDS_CODES.A,
      grid: mockGrid,
      obstaclesCoordinates: mockObstaclesCoordinates,
    });
  });

  it('Should fire new keyboardMove on press D', () => {
    const { getByLabelText } = setup();
    fireEvent.keyDown(getByLabelText('Grid'), { key: 'D', code: 'KeyD' });
    expect(roverNewKeyboardMove).toHaveBeenCalledTimes(1);
    expect(roverNewKeyboardMove).toHaveBeenCalledWith({
      code: KEYBOARDS_CODES.D,
      grid: mockGrid,
      obstaclesCoordinates: mockObstaclesCoordinates,
    });
  });

  it('Should fire new keyboardMove on press W', () => {
    const { getByLabelText } = setup();
    fireEvent.keyDown(getByLabelText('Grid'), { key: 'W', code: 'KeyW' });
    expect(roverNewKeyboardMove).toHaveBeenCalledTimes(1);
    expect(roverNewKeyboardMove).toHaveBeenCalledWith({
      code: KEYBOARDS_CODES.W,
      grid: mockGrid,
      obstaclesCoordinates: mockObstaclesCoordinates,
    });
  });

  it('Should not fire new keyboardMove on press S', () => {
    const { getByLabelText } = setup();
    fireEvent.keyDown(getByLabelText('Grid'), { key: 'S', code: 'KeyS' });
    expect(roverNewKeyboardMove).toHaveBeenCalledTimes(0);
  });

  it('Should clear data when unmounting the component', () => {
    const { unmount } = setup();
    unmount();
    expect(roverClear).toHaveBeenCalledTimes(1);
  });
});
