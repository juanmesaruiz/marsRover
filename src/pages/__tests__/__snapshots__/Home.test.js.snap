// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home test suite Should match snapshot 1`] = `
.c0 {
  padding: 50px;
  text-align: center;
}

.c0 .main-container {
  margin-top: 20px;
}

<div>
  <div
    class="c0"
  >
    <header>
      <code
        aria-label="Title"
        name="Title"
        props="[object Object]"
      />
    </header>
    <section>
      <code
        aria-label="Description"
        name="Description"
        props="[object Object]"
      />
    </section>
    <main
      class="main-container"
    >
      <code
        aria-label="Grid"
        name="Grid"
        props="[object Object]"
      />
      <code
        aria-label="LogPosition"
        name="LogPosition"
        props="[object Object]"
      />
      <code
        aria-label="Actions"
        name="Actions"
        props="[object Object]"
      />
    </main>
  </div>
</div>
`;
